From: Benjamin Kaduk <kaduk@mit.edu>
Date: Thu, 23 Oct 2014 17:28:43 -0400
Subject: Add dummy -exit command for afsd to do nothing

---
 src/afsd/afsd.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/src/afsd/afsd.c b/src/afsd/afsd.c
index 0976099..fcf3988 100644
--- a/src/afsd/afsd.c
+++ b/src/afsd/afsd.c
@@ -362,6 +362,7 @@ enum optionsList {
     OPT_dynrootsparse,
     OPT_rxmaxfrags,
     OPT_inumcalc,
+    OPT_exit,
 };
 
 #ifdef MACOS_EVENT_HANDLING
@@ -1905,6 +1906,10 @@ CheckOptions(struct cmd_syndesc *as)
     if (cmd_OptionPresent(as, OPT_inumcalc)) {
 	cmd_OptionAsString(as, OPT_inumcalc, &inumcalc);
     }
+    if (as->parms[39].items) {
+	/* -exit */
+	exit(0);
+    }
 
     /* parse cacheinfo file if this is a diskcache */
     if (ParseCacheInfoFile()) {
@@ -2624,6 +2629,8 @@ afsd_init(void)
 			"send/receive per Rx packet");
     cmd_AddParmAtOffset(ts, OPT_inumcalc, "-inumcalc", CMD_SINGLE, CMD_OPTIONAL,
 			"Set inode number calculation method");
+    cmd_AddParmAtOffset(ts, OPT_exit, "-exit", CMD_FLAG, CMD_OPTIONAL,
+                        "exit immediately");
 }
 
 /**
